README.txt
This file is used to list a few config items and recommendation. Also some basic Praeda syntax 


Required perl modules:
LWP::Simple
LWP::UserAgent
HTML::TagParser
URI::Fetch
HTTP::Cookies
IO::Socket
HTML::TableExtract
Getopt::Std
Net::SSL
Net::SNMP
NetAddr::IP;

Quick CPAN command to make this happen:
cpan -i LWP::Simple LWP::UserAgent HTML::TagParser URI::Fetch HTTP::Cookies IO::Socket HTML::TableExtract Getopt::Std  Net::SSL Net::SNMP NetAddr::IP


-----------------------------------------------------------------
Note: If running on Windows host and having issues install Net:SSL with cpan. 
1.Change line 47 in praeda.pl from “use Net::SSL; to use Net::SSLeay;
2. Run following install for Net-SSLeay.ppd:
	ppm install http://www.sisyphusion.tk/ppm/Net-SSLeay.ppd
-----------------------------------------------------------------


set root of praeda install in praeda.pl under
my $dirpath =".";



Praeda version 0.02.3.117b syntax:
first part of version  0.02.3 represent the core product version the Last 3 digits represent the number of product versions tested.

PRAEDA OPTIONS:
-g GNMAP_FILE
-n CIDR or CIDR_FILE 
-t TARGET_FILE
-p TCP_PORT
-j PTOJECT_NAME
-l OUTPUT_LOG_FILE
-S SSL

GNMAP_FILE = This is a .gnmap file output by a nmap scan.
CIDR & CIDR_FILE = Subnet CIDR "192.168.1.0/24" or file containing list of CIDRs
TARGET_FILE = List of IP addresses or Host names to enumerated
TCP_PORT = port address of targets to scan " At present only one port can be specified. This is expected to be modified in future version"
PROJECT_NAME = the name for this project. This will create a folder under the folder where Praeda was executed to contain logs and export info.
OUTPUT_LOG_FILE = name of log file for data output


SYNTAX FOR GNMAP FILE INPUT:
praeda.pl -g GNMAP_FILE -j PROJECT_NAME -l OUTPUT_LOG_FILE

SYNTAX FOR IP  CIDR/CIDR FILE LIST:
praeda.pl -t CIDR or CIDR_FILE -p TCP_PORT -j PROJECT_NAME -l OUTPUT_LOG_FILE -s SSL


SYNTAX FOR IP TARGET FILE LIST:
praeda.pl -t TARGET_FILE -p TCP_PORT -j PROJECT_NAME -l OUTPUT_LOG_FILE -s SSL 
 

Examples:

./praeda.pl -g scan1.gnmap -j acmewidget -l results

./praeda.pl  -n 10.10.10.0/24 -p 80  -j project1 -l data-file

./praeda.pl  -n cidrs.txt -p 80  -j project1 -l data-file

./praeda.pl  -t target.txt -p 80  -j project1 -l data-file

./praeda.pl  -t target.txt -p 443  -j project1 -l data-file -s SSL


NOTE: The -n option has a limit on networks of 65535 so cidr mask less the /16 will error out with "netmask error: overrange". This is a limitation of the module NetAddr::IP.


The results will create a folder called project1 and save all information in that folder. Also this will write out the following data.
targetdata.txt  : This is the parsed results of .gnmap file
$LOGFILE-WebHost.txt : This is an output of all webservers querried listing IP:PORT:TITLE:SERVER
$LOGFILE.log : This file will contain the results of the modules executed.
RAW extract data including: Clones, Backups, Address Books ect...

****WARNING**** 
Also insure that your local firewall is turned off. Certain modules that require connection back to host system for the module to run correctly.


CHANGELOG.txt
May 13, 2011  added line to praeda.pl to replace non-printable characters with spaces in the extracted title from targeted systems during finger printing process "Deral Heiland"

June 4, 2011  added Modules MP0014.pl for Sharps LDAP capture leveraging a connect back attack

June 10, 2011 added Modules MP0015.pl for Xerox Workcentra Address Book  Download


June 25, 2011  re-enabled SSL by adding code to enable -ssl and  and ignore invalid certs supplied by Chris Teodorski  "Deral Heiland"

July 21, 2011  added Module MP0016.pl for Lanier (Ricoh) MP C5000/LD550C and LD335c username enumeration supplied by Chris Teadorski

July 21, 2011  modified praeda.pl added use HTML::TableExtract also modified @,$,MY to fix use strict; complaints

Aug 2, 2011  added Module MP0017.pl for Lexmark extract settingfile.ucf which contains configuration information including plan text passwords "Deral Heiland"

Sep 2, 2011  Added 2 new modules  MA0008.pl which test for Spectrum24 Access point and test fordefault creds, snmp strings and wep keys. Moldule MA0009.pl is for APC Network Managment card and test for default creds and snmp strings. both modules supplied by Chris Teadorski

Oct 20, 2011 Disabled modules MP0016.pl because of issues causing failure, on drawing board for more testing. Also a couple simple bug fix in MA0008.pl & MA0009.pl had exit set if default password was not found causing Praeda to exit and not complete properly.

Nov 07, 2011 made a few modification to MP0008.pl and MP0015.pl to support Xerox workcentra 57XX series MFP devices changes support clone and address book downloads from 57XX devices. Also  updated datalist with 57XX jobs

Jan 10, 2012 Added option switches to praeda.pl. Added feature to allow the use of NMAP .gnmap output as target input to praeda.pl (Version release 0.02.0.0b). Also modified all modules text output to state "Attempting". Also add "404 Not Found" check to MP0005.pl.

Mar 1, 2012 Added two new modules, MP00T2.pl "HP CM3530" and MP00T1.pl "HP Officejet 6500n-z". Modified Praeda.pl to allow a compare of =~ when validating finger printer server type. This allows for support of changing server type information by comparing part of string to match versus 100% match.

Mar 8, 2012 added a 403 Forbidden check of returned title page in MP0008.pl. In case basic auth fails. This prevent the creation of bad cloning.dlm files.

Mar 22, 2012 added 3 new appliance modules. 2 check for default creds MA0011 "Digi" & MA0013 "Integrated Remote Access Controller 6 'IDRAC6'"  and one that checks for snmp string MA0012 "EtherPath".

June 15, 2012 added 1 new appliance module MA0014.pl. extracts the snmp string setting used by a Pillar Axiom Performace Statistics Monitor"

Sep 6, 2012 added 1 new appliance module MA0015.pl check for default creds on a CoyotePoint Equalizer device

Oct 12, 2012 added new module MP0019.pl which validated defaut creds for Konica Minolta 1690MFM and 4690MFP devices and also attempts to retrieve LDAP, POP3 and SMTP authentication credentials.

Feb 1, 2013 added support to pull device finger print data via snmp queries. This data is recorded in the5th column in the output log files. Also added new module MP0020.pl that leverages the new SNMP to identify a Konica C452 and extract SMBMODE information including Host, Username, Password, and folder name.

Feb 2, 2013 updated MP0001 and MP0002 HP color usage logs extract to pars out duplicates.

Apr 17, 2013 added 1 new appliance module MA0016.pl check  for default creds and extraxt snmp settings on Emerson Network Power IntelliSlot Web Card.

May 5, 2013 added 1 new module MP0021.pl checks default creds, enable password export withing address book export, disables encrypting and exports address books with plan text passwords on Canon IR-ADV MFP devices.

May 17, 2013 added 1 new module MA0017.pl which checks for default creds and snmp community strings on Tripp Lite Power Alert Device. also updated version number to 0.02.1.072b. and added a bug fix to MA0016.pl.

May 31, 2013 added 1 new module MA0018.pl which checks for default creds on a Binatone ADSL Wireless Modem Router DT 845W

June 30, 2013 Modified module MA0018.pl to check for a number of rebranded ADSL Wireless Modem Routers. Original finger print is ambiguous so will check for different password of password and admin. Also modified several modules to list ip address on same line as SUCCESS when doing default password checks. Make for easier parsing of output for info.

July 19, 2013 Added Module MA0019.pl to check for default creds and pull SNMP community strings from Pelco Sarix Camera.

August 18, 2013 bug fix issue with processing HTTPS in praeda.pl when using gnmap data in input.

August 29, 2013 Added new support modules for extraction of devices admin passwords for following Workcentre,s 5638, 5645, 5655, 5735, 5740, 5745, 5755, 7775. Also added support for extraction of SMB creds from address books on KONICA MINOLTA bizhub C220 & 223.

September 16, 2013 added support for a KONICA MINOLTA bizhub 423 & 552

October 17, 2013 added functionality to scan via CIDR (ie 192.168.1.0/24) or via a list of CIDRs Version upgrade to 2.2.0.089b

January 25, 2014  added check and Title field padding - found strange issues when dealing with camparing enumerated Title pages and array containing Null data

Feburary 13, 2014  added two new Canon devices to the data_list iR-ADV C5045 & 4051

May 17, 2014 added new support modules for extracting username, passwords, WPA PSK and WEP keys for Ubee, Ambit and Netopia modems.

Aug 03, 2014 added new support for arris wifi modem SBG6580. Also disabled Riverbed module (MA0001)high false positives, and added a few more test finger prints for older APCs.


praeda.pl
#!/usr/bin/perl
#
# praeda.pl
# praeda [robber, plunderer]. 
#   
#
# PRAEDA version 0.02.3.118b
######################################################
#                    PRAEDA                          #
#               Copyright (C) 2014                   #
#              Deral 'percX' Heiland                 #
#                  @percent_x	                     #
######################################################
#
#  This program is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License
#  as published by the Free Software Foundation; version 2
#  of the License only.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#
# Contact Information:
#       dh@layereddefense.com
#######################################################################
# See the README.txt and/or help files for more information on how to use & config.
# See the LICENSE.txt file for more information on the License PRAEDO is distributed under.
#
# This program is intended for use in an authorized manner only, and the author
# can not be held liable for anything done with this program, code, or items discovered
# with this program's use.
#######################################################################
use strict;
use LWP::Simple;
use LWP::UserAgent;
use HTML::TagParser;
use URI::Fetch;
use HTTP::Cookies;
use IO::Socket ();
use HTML::TableExtract;
use Getopt::Std;
use Net::SSL;
use Net::SNMP;
use NetAddr::IP;

# -- Set Variables ---------------
my $dirpath =".";
my $DataEntry ="";
my $number = "";
my $title = "";
my $server = "";
my $job = "job";  
my $web = "";
my $SOCKET_IS_UP = 0;
my $SOCKET_IS_DOWN = 1;
my $REQUEST_TIMEOUT = 5;
my $WAIT_FOR_RST = 1;
my $TARGET = "";
my $PORTS = "";
my $OUTPUT = "";
my $NAME = "";
my $LOGFILE = "";
my $FILE = "";
my $CIDRFIL = "";
my $data3 = "Q";

# -- Set Options ---------------
my %options=();
getopts("g:n:t:p:j:l:s:",\%options);

#set option exclusions and messages
if (($options{g} && ($options{t} || $options{p})) || ($options{g} && ($options{n} || $options{p})) || ($options{g} && $options{n})) {
  print"-g and -t or -p options are not allowed at same time\n";
  print"The correct options syntax are:\n";
  print"For GNMAP input:  praeda.pl -g GNMAP_FILE -j PROJECT_NAME -l OUTPUT_LOG_FILE\n";
  print"For target input: praeda.pl -t TARGET_FILE -p TCP_PORT -j PROJECT_NAME -l OUTPUT_LOG_FILE -s SSL \n";
  print"For CIDR input:   praeda.pl -n CIDR or CIDR_FILE -p TCP_PORT -j PROJECT_NAME -l OUTPUT_LOG_FILE -s SSL \n";
  exit;
}
elsif ($options{g} && (!$options{j} || !$options{l})) {
  print"Options -j and -l are both required when using option -g\n";
  print"The correct options syntax for using gnmap as input is:\n";
  print"praeda.pl -g GNMAP_FILE -j PROJECT_NAME -l OUTPUT_LOG_FILE\n";
  exit;  
}
elsif ($options{n} && (!$options{p} || !$options{j} || !$options{l})) {
  print"Options -p, -j and -l are all required when using option -n\n";
  print"The correct options syntax for using network CIDR or CIDR list file as input is:\n";
  print"For CIDR input: praeda.pl -n CIDR or CIDR_FILE -p TCP_PORT -j PROJECT_NAME -l OUTPUT_LOG_FILE -s SSL \n";  
  exit;
}
elsif ($options{t} && (!$options{p} || !$options{j} || !$options{l})) {
  print"Options -p, -j and -l are all required when using option -t\n";
  print"The correct options syntax for using target ip list file as input is:\n";
  print"praeda.pl -t TARGET_FILE -p TCP_PORT -j PROJECT_NAME -l OUTPUT_LOG_FILE -s SSL \n";
  exit;  
}
elsif (!$options{g} && !$options{t} && !$options{n} ) {
  print"Required options are missing\n";
  print"The correct options syntax are:\n";
  print"For GNMAP input:  praeda.pl -g GNMAP_FILE -j PROJECT_NAME -l OUTPUT_LOG_FILE\n";
  print"For target input: praeda.pl -t TARGET_FILE -p TCP_PORT -j PROJECT_NAME -l OUTPUT_LOG_FILE -s SSL \n";
  print"For CIDR input:   praeda.pl -n CIDR or CIDR_FILE -p TCP_PORT -j PROJECT_NAME -l OUTPUT_LOG_FILE -s SSL \n";
  exit;  
}

# -- Import data_file ---------------
my $data_file="$dirpath/data/data_list";
open(DAT, $data_file) || die("Could not open file!");
my @raw_data=<DAT>;
close(DAT);

# -- Enable -ssl ---------------
if (($options{s} eq "ssl" ) || ($options{s} eq "SSL"))
        {
	$web = 's';
        }

# -- Set Ingnore Invalid Certs ---------------
$ENV{'PERL_LWP_SSL_VERIFY_HOSTNAME'}=0;


# -- Setup Browser ---------------
my $browser = LWP::UserAgent->new;
$browser->cookie_jar({});
$browser->timeout(15);

# -- Create Project Folder ---------------
mkdir "$options{j}", 0775 unless -d "$options{j}";


# -- Setup/Call GNMAP Routine ---------------
if ($options{g}){
my $GNMAPFILE = $options{g};
$OUTPUT = $options{j};
$NAME = $options{l};
&gnmap_parse($options{g}, $options{j});
}

# -- Setup/Call CIDR Routine ---------------
if ($options{n}){
$CIDRFIL = $options{n};
$OUTPUT = $options{j};
$NAME = $options{l};
&cidr_parse($options{n}, $options{j});
}

# -- Set Variables from Options --------------- 
if ($options{t})
  {
  $FILE = "$options{t}";
  $PORTS = "$options{p}";
  }
elsif ($options{n})
 {
  $FILE = "./$options{j}/targetdata.txt"; 
  $PORTS = "$options{p}";
 }
else
  {
  $FILE = "./$options{j}/targetdata.txt";
  }
$OUTPUT = "$options{j}";
$LOGFILE = "$options{l}";



# -- Read Target Input ---------------
open (FILE, "$FILE") || die("Unable to open: $FILE $!");

while (<FILE>)
	{
	if ($options{g})
             {
              chomp $TARGET;
              ($TARGET, $PORTS, my $N) = split(/:/);
                if (($N =~ m/https/i) || ($N =~ m/ssl/i)){ $web = "s";}
		else {$web ="";}
              }
        elsif ($options{n})
             {
              $TARGET = $_; 
              chomp $TARGET;
             }
            
        else 
             {
              $TARGET = $_;
              chomp $TARGET;
             }

# -- Call Port Check Routine --------------- 
	my ( $status ) = 
	check_port( $TARGET, $PORTS );


if ( $status == $SOCKET_IS_DOWN )
     {
      # open(WEBFILE, ">>./$OUTPUT/$LOGFILE-WebHost.txt") || die("Failed to open  Output file $LOGFILE-webhost.txt \n");
      print "$TARGET:$PORTS:NO ANSWER RETURNED\n";
      # print WEBFILE "$TARGET:$PORTS:NO ANSWER RETURNED\n";
      # close(WEBFILE);
     }
# -- Target Enumeration Section ---------------
else
     {
     my $html = $browser->get("http$web://$TARGET:$PORTS/");
     my $data1 = $html->header("Title");

     $data1 =~ s/[^[:print:]]/ /g;  # replace nonprintable characters with spaces "added May 13 2011 percX"
     if ($data1 eq "") # padding Title field with space to avoid NULL array issues "added Jan 26 2014 percX"
       {
        $data1 =~ s/^/ /g;
       }

     my $data2 = $html->header("Server");
     print "$TARGET:$PORTS:$data1:$data2\n";

     # SNMP Device Information Check 
     (my $session,my $error) = Net::SNMP->session(Hostname => $TARGET, Community => "public",timeout => 1);
     my $result = $session->get_request("1.3.6.1.2.1.1.1.0");
     $session->close;
     $data3 = $result->{"1.3.6.1.2.1.1.1.0"}; 

     open(WEBFILE, ">>./$OUTPUT/$LOGFILE-WebHost.txt") || die("Failed to open  Output file $LOGFILE-webhost.txt \n");
     print WEBFILE "$TARGET:$PORTS:$data1:$data2:$data3\n";
     close(WEBFILE);                
            foreach $DataEntry (@raw_data)
               {
                chomp($DataEntry);
                my @values=split(/\|/,$DataEntry);
                  if ((($data1 eq $values[1]) && ($data2 =~ $values[2])) || (($data3 =~ $values[1]) && ($values[2] eq "SNMP")))
                     {
 		       my $num = $#values + 1;
                       for (my $i=3;$i<$num;$i++)
			{
                          if ($values[$i] eq ''){}
                          else
                             {
                             open(OUTFILE, ">>./$OUTPUT/$LOGFILE.log") || die("Failed to open  Output file $LOGFILE.log \n");
                             print OUTFILE "\n$TARGET:$PORTS:$data1:$data2:$data3\n";
			     close (OUTFILE);
			     $job = $values[$i];	
			     require "$dirpath/jobs/$values[$i].pl";
                             my $printpwn = $job->$job($TARGET,$PORTS,$web,$OUTPUT,$LOGFILE,$data1,$data3);
			     print "\n";
                             } 
			}

                                  
                    } 
                }
     }

}

#-----------------------------------------------subroutines------------------------------------------------------#

# -- TCP Port Check Routine ---------------
 sub check_port {
  my ( $TARGET, $PORTS ) = @_;
  my $status = $SOCKET_IS_DOWN;
  my $socket = IO::Socket::INET->new(
    PeerAddr => $TARGET,
    PeerPort => $PORTS,
    Timeout => $REQUEST_TIMEOUT 
    ); 


 if ( defined $socket ) 
     {
      # sleep $WAIT_FOR_RST; 
      my $is_connected = $socket->connected;

         if ( defined $is_connected ) 
            {
            $status = $SOCKET_IS_UP;
            } 
     }
return ( $status);
}

# -- GNMAP Parse and Save Routine ---------------
sub gnmap_parse {
        my ( $GNMAPFILE, $OUTPUT ) = @_;
	open(MYINPUTFILE, "$GNMAPFILE") || die("Failed to open gnmap file $GNMAPFILE \n");
	unlink("./$OUTPUT/targetdata.txt");

while(<MYINPUTFILE>)
{
    my @portspec = ();
    my @gnmap = ();
    next if ( /^\#/ );                                 		# skip comments
    next if ( /^\s*$/ );                			# skip blank lines
    next if ( /\tStatus: Down/ );       			# skip down hosts
    chomp;

    # clean up the entry
    s!host:\s+!!i;                      			# Host:
    s!\s+\(\)\s+!:!g;                   			# missing hostname
    s!^(\d+\.\d+\.\d+\.\d+)\s+(\([\w\.\-\_]+\))\s*!$1:$2!; 	# IP (hostname) to IP:(hostname)
    s!ports: !:!i;                      			# "Ports"
        s!|.*!!;                        			# trailing garbage
    s!\t+Ign.*!!;                       			# "Ignored State"

    # extract IP addresses and ports found
    my @row = split (/:/, $_, 3);
    my @portspec = split (/,/, $row[2]);
    
    #save only ports marked "open" and contain http
    open(OUTFILE, ">>./$OUTPUT/targetdata.txt") || die("Failed to open  Output file targetdata.txt \n");
    for my $portspec ( @portspec )
	{
	   my @fields = split ( /\//, $portspec );
	   next unless (( $fields[1] =~ /open/i ) && ($fields[4] =~ m/http/i));
	   my $p = "$row[0]:$fields[0]:$fields[4]";
	   $p =~ s/ //;
           print OUTFILE "$p\n";
	}
    close (OUTFILE);
}


# -- CIDR Parse Routine ---------------
sub cidr_parse {
  my ( $GNMAPFILE, $OUTPUT ) = @_;
  unlink("./$OUTPUT/targetdata.txt");
  open(OUTFILE, ">>./$OUTPUT/targetdata.txt") || die("Failed to open  Output file targetdata.txt \n");

   # extract ip addresses from CIDR or CIDR file and output to targetdata.txt file
   if ( -e $CIDRFIL )
      {
       open(HAND, $CIDRFIL) || die("Unable to open: $CIDRFIL $!");
       my @cidr=<HAND>;
       close(HAND);
       for my $cidr( @cidr )
          {
           my $n = NetAddr::IP->new( $cidr );
           for my $ip( @{$n->hostenumref} )
              {
               print OUTFILE $ip->addr, "\n";
              }
          }
      }

   else
      {
       my $cidr = $CIDRFIL;
       my $n = NetAddr::IP->new( $cidr );
       for my $ip( @{$n->hostenumref} )
          {
           my $errText = $!;
           chomp($errText);
           print OUTFILE $ip->addr, "\n";
          }
       }
        close (OUTFILE);
       }
}
